drivers/rtc/rtc_mc146818: Fix RTC driver for QEMU boards

bjarki-andreasen opened the pull request:
This commit fixes the driver, allowing it to pass
the RTC API test suite for set/get time, alarms and the update callback (the hw does not support calibration)

This requires an additional flag to the HPET module which disables legacy IRQs, allowing the RTCs IRQ to be raised.

Furthermore, the commit cleans up the RTC in the devicetree for the IA32 dtsi and the qemu_x64 dts files, and updates the yaml files for QEMU to indicate the support for the RTC subsystem.

Fixes #59448

bjarki-trackunit commented:
Proof of it working @cfriedt  :D

```
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: qemu32,+nx,+pae
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build zephyr-v3.4.0-974-g7fba7d395750 ***
Running TESTSUITE rtc_api
===================================================================
START - test_alarm
 PASS - test_alarm in 26.041 seconds
===================================================================
START - test_alarm_callback
 PASS - test_alarm_callback in 26.040 seconds
===================================================================
START - test_set_get_time
 PASS - test_set_get_time in 0.001 seconds
===================================================================
START - test_time_counting
 PASS - test_time_counting in 9.689 seconds
===================================================================
START - test_update_callback
 PASS - test_update_callback in 15.020 seconds
===================================================================
START - test_y2k
 PASS - test_y2k in 2.010 seconds
===================================================================
TESTSUITE rtc_api succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [rtc_api]: pass = 6, fail = 0, skip = 0, total = 6 duration = 78.801 seconds
 - PASS - [rtc_api.test_alarm] duration = 26.041 seconds
 - PASS - [rtc_api.test_alarm_callback] duration = 26.040 seconds
 - PASS - [rtc_api.test_set_get_time] duration = 0.001 seconds
 - PASS - [rtc_api.test_time_counting] duration = 9.689 seconds
 - PASS - [rtc_api.test_update_callback] duration = 15.020 seconds
 - PASS - [rtc_api.test_y2k] duration = 2.010 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

cfriedt commented on drivers/rtc/rtc_mc146818.c:
```diff
@@ -545,26 +463,29 @@ static int rtc_mc146818_update_set_callback(const struct device *dev,
 	struct rtc_mc146818_data * const dev_data = dev->data;
+	uint8_t regc;
 
-	ARG_UNUSED(dev_data);
```
Undelete this line

bjarki-trackunit replied:
Tiny fix applied

cfriedt reviewed (commented):
Tiny fix

bjarki-trackunit commented:
@dcpleung @andyross 

I need help solving a timeout occurring for a library named p4workq, which times out if `no-legacy-irq;` is used for the HPET for the QEMU boards. The test timing out is a stress test, but it seems it is a bug rather than an increased workload causing the timeout.

I am running the test using `west twister -p qemu_x86_tiny -T zephyr/tests/lib/p4workq`

The stress test has a define `#define MAX_EVENTS` which scales the amount of work which needs to be done. I ran the test for a few different values:

100 -> 13.27 seconds
300 -> 13.10 seconds
400 -> 13.37 seconds
410 -> timed out after 64 seconds...

I am enabling `no-legacy-irq;` to allow the QEMU q35 RTC's IRQ to fire.

I need help to move forward :)

cfriedt commented on drivers/rtc/rtc_mc146818.c:
```diff
@@ -51,7 +51,6 @@
 #define RTC_UIP		RTC_REG_A
 #define RTC_DATA	RTC_REG_B
 #define RTC_FLAG	RTC_REG_C
-#define RTC_ALARM_MDAY	RTC_REG_D
```
How much of this file actually needed to be changed in order for the fix to work?

At a glance, it seems that most of the changes here are cosmetic (formatting, name changes, ...). Is it possible to reduce the change to only what is required for the fix?

bjarki-trackunit replied:
These changes are the bare minimum required to get the driver to an acceptable state. The driver still needs a once over to clean it up, which I agree is not within the scope of this PR

cfriedt commented on boards/x86/qemu_x86/qemu_x86.dts:
```diff
@@ -132,6 +132,10 @@
 	status = "okay";
 };
 
+&rtc {
```
This could likely be moved to the RTC testsuite in a board overlay, since it is causing unrelated things to fail, maybe with a comment saying why it's necessary.

bjarki-trackunit replied:
I will move it to the test suite, and make sure the RTC tests are built

cfriedt commented on dts/x86/intel/ia32.dtsi:
```diff
@@ -64,6 +64,7 @@
 			reg = <0xfed00000 0x400>;
 			interrupts = <2 IRQ_TYPE_FIXED_EDGE_RISING 4>;
 			interrupt-parent = <&intc>;
+			no-legacy-irq;
```
This can likely also be added in the testsuite in a board overlay, maybe with a descriptive comment, in order to avoid breaking unrelated things.

bjarki-trackunit replied:
This will also go to the test suite

cfriedt commented on dts/x86/intel/ia32.dtsi:
```diff
@@ -75,8 +76,7 @@
 			alarms-count = <1>;
 
-			status = "okay";
+			status = "disabled";
```
This can likely also be added in the testsuite in a board overlay, maybe with a descriptive comment, in order to avoid breaking unrelated things.

bjarki-trackunit replied:
I don't think this should be in the test suite, I don't know what criteria we use to determine if a node is enabled in the dtsi file, but for an RTC, I would expect it to be disabled by default

bjarki-trackunit replied:
Removed this change from the PR

cfriedt reviewed (changes requested):
It looks like a lot of changes in the driver are potentially unrelated to the actual fix.

I would suggest separating any cosmetic changes, and functional changes to separate commits.

Other than that, I think to avoid breaking other things (libp4work or whatever it is), it might be better to make the necessary DT adjustments in a board overlay in the testsuite.

Also, is it possible to ensure the driver is built and tested in CI? Do we have it in a build all test somewhere?

Commit:
drivers/rtc/mc146818: Patch MC146818 driver

The MC146818 driver was not properly initialized
by the driver, interrupts where not handled correctly,
and the alarm feature was not implemented properly.

This commit fixes these issues, while removing some
code which became redundant as the MC146818 driver
was patched.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
tests/drivers/rtc: Add MC146818 driver to test suite

This commit updates the qemu_x86 board's yaml file to
indicate its support for the RTC subsystem. Board overlay
and conf for the qemu_x86 has been added to the RTC test
suite to enable the MC146818 and its dependencies.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
tests/drivers/rtc/rtc_api: Patch 64-bit arch support for test

The update callback test had a bug which prevented it from
running on 64-bit architectures. This patch makes the test
agnostic to 64-bit and 32-bit architectures.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

cfriedt reviewed (dismissed):
Tests are passing 🙌

cfriedt commented:
I realize I just approved this, but it looks like a few of the RTC kconfig options are missing from `tests/drivers/rtc/rtc_api/prj.conf`

From what I understand all of the rtcs exercised in this testsuite support all of the options, so might be better to add them there as well.

bjarki-trackunit commented:
@cfriedt

Only the mandatory tests from the test suite are run by default, which is why the features RTC_ALARM and RTC_UPDATE and RTC_CALIBRATE are not included in the main .prj Features are selected in the board overlays. The MC146818 only supports RTC_ALARM and RTC_UPDATE, not RTC_CALIBRATE :)

cfriedt commented:
> @cfriedt
> 
> Only the mandatory tests from the test suite are run by default, which is why the features RTC_ALARM and RTC_UPDATE and RTC_CALIBRATE are not included in the main .prj Features are selected in the board overlays. The MC146818 only supports RTC_ALARM and RTC_UPDATE for example, not RTC_CALIBRATE :) 

Sure, but just remember we need code paths tested, even if they are disabled by default

akanisetti commented on drivers/rtc/rtc_mc146818.c:
```diff
@@ -583,26 +506,30 @@ struct rtc_driver_api rtc_mc146818_driver_api = {
+#define RTC_MC146818_INIT_FN_DEFINE(n)						\
+	static int rtc_mc146818_init##n(const struct device *dev)		\
+	{									\
+		rtc_write(RTC_REG_A, 0);					\
```
Clearing RTC_REG_A is blocking RTC time update on real hardware.

bjarki-trackunit replied:
I see, the QEMU board has no oscillator connected to the RTC (the clock therefore runs regardless of divider settings), my theory is that your RTC is connected to an oscillator. Could you check if setting the DV2, DV1 and DV0 bits according to the RTC's clock input? If that works, I will expand the dts and driver with input clock support :)

akanisetti replied:
Yes, setting DV2-0 according to RTC clock allows it run. Freq on my hardware is 32.786KHz so DV sequence is 0-1-0.

bjarki-trackunit replied:
I have added input clock frequency configuration to the driver and devicetree (See [commit](https://github.com/zephyrproject-rtos/zephyr/pull/60321/commits/f33e973278b587cd764916181a1f34ffc9a51ad7)). Please test it out :)

akanisetti replied:
It is working👍

bjarki-trackunit replied:
Awesome :)

Commit:
drivers/rtc/rtc_mc146818.c: Add input clock config to driver

This commit adds input clock selection to the RTC driver. This
is required to allow for the real hardware to operate. The
QEMU emulated hardware ignores the input clock settings.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
tests/drivers/rtc: Add qemu_x86_64 board to test suite

This commit adds overlay and conf for the qemu_x86_64 board
to the RTC API test suite, and adds support for the RTC
subsystem to the qemu_x86_64's yaml file.

The commit also specifies integration platforms for the
RTC API test suite, since the qemu_x86_64 board runs in
real-time, causing it to time out if the test suite runs
for it.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

cfriedt commented:
is @akanisetti able to review / approve as well?

cfriedt commented:
@bjarki-trackunit - you might need to find another reviewer with approve privileges.

bjarki-trackunit commented:
@henrikbrixandersen @erwango Could you please review this PR? :)

nordicjm commented on drivers/rtc/rtc_mc146818.c:
```diff
@@ -113,6 +113,11 @@
 #define MAX_YEAR_DIFF	99 /* YEAR - 1999 */
 
+/* Input clock frequency mapped to divider bits */
+#define RTC_IN_CLK_DIV_BITS_4194304 (0b000 << 4)
+#define RTC_IN_CLK_DIV_BITS_1048576 (0b001 << 4)
+#define RTC_IN_CLK_DIV_BITS_32768   (0b010 << 4)
```
not acceptable to use C23 features

bjarki-trackunit commented:
@nordicjm @cfriedt Could you kindly re-review? :)

cfriedt commented on tests/drivers/rtc/rtc_api/boards/qemu_x86_64.overlay:
```diff
@@ -0,0 +1,18 @@
+/*
+ * The RTC IRQ is not routed to the IOAPIC if the legacy
+ * IRQ bit is set. The IRQ is required for alarm
+ * operation and the update callbadk.
```
Nit: spelling mistake

cfriedt reviewed (dismissed):
One tiny spelling mistake - I'll re-approve when fixed.

bjarki-trackunit commented:
@cfriedt pretty please :D
